### DISCUSSION TITLE
Use/share global typescript declarations in apps and libs

### DISCUSSION THREAD

**Discussion by stopyransky** [nrwl/nx#8838] [General]
I have a problem in using global typescript declarations among nx libs.

Below setup works fine for all apps:

I have `index.d.ts` in a root directory:

```
// /index.d.ts
declare global {
  namespace MyNamespace {
     interface MyInterface { ... }
  }
}
```

In `tsconfig.base.json` I have added `"esModuleInterop": true,` and ` "typeRoots": ["index.d.ts"]` in `compilerOptions`.
Then in each `/apps/my-app/tsconfg.json` I added   ` "../../index.d.ts" ` in `include` array.

This allows me to use global declarations in apps without the need to import anything.

`const myVar: MyNamespace.MyInterface = ...`

I tried to do the same for libs but it does not work. 

There are three tsconfig.jsons in lib directory:
 - tsconfig.json (main)
 - tsconfig.lib.json (lib)
 - tsconfig.spec.json (spec)
 and if i use a storybook for a lib then there is fourth `/.storybook/tsconfig.json` (storybook).

lib, spec and storybook configs extend main config.

There is error in the file:

```
Referenced project '/.../libs/shared-one/tsconfig.lib.json' must have setting "composite": true.
```

if i set composite to true for lib, spec, storybook configs then I get following error there:

```
Composite projects may not disable declaration emit.
```

What is wrong? How do I share global type declarations among nx libs? I hoped this will work out of the box.

example of tsconfig.lib.json (lib)
```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "../../dist/out-tsc",
    "types": ["node"]
  },
  "files": [
    "../../node_modules/@nrwl/react/typings/cssmodule.d.ts",
    "../../node_modules/@nrwl/react/typings/image.d.ts"
  ],
  "exclude": [
    "**/*.spec.ts",
    "**/*.spec.tsx",
    "**/*.stories.ts",
    "**/*.stories.js",
    "**/*.stories.jsx",
    "**/*.stories.tsx"
  ],
  "include": ["**/*.js", "**/*.jsx", "**/*.ts", "**/*.tsx"]
}
```

/libs/my-lib tsconfig.json (main)

```
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "jsx": "react-jsx",
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
  },
  "files": [],
  "include": ["../../index.d.ts"],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    },
    {
      "path": "./.storybook/tsconfig.json"
    }
  ]
}
```

// tsconfig.base.json: (base)
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "rootDir": ".",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "importHelpers": true,
    "target": "es2015",
    "module": "esnext",
    "lib": ["es2017", "dom"],
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "baseUrl": ".",
    "paths": {
      "@my-org/shared-one": ["libs/shared-one/src/index.ts"],
      "@my-org/shared-two": ["libs/shared-two/src/index.ts"]
    }
  },
  "exclude": ["node_modules", "tmp"],
  "include": ["index.d.ts"]
}
```

**Comment by ALFmachine** [comment]
i was able to resolve this in my code base by adding custom types to the tsconfig.base.json after declaring them in the typeRoots

tsconfig.base.json
```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "rootDir": ".",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "target": "es2015",
    "module": "esnext",
    "lib": ["es2017", "dom"],
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "baseUrl": ".",
    "paths": {
      "@prisma-client": ["packages/libs/prisma-client/src/index.ts"],
      "@ui/meeting": ["packages/libs/ui/meeting/src/index.ts"],
      "@ui/shared": ["packages/libs/ui/shared/src/index.ts"],
      "@ui/util": ["packages/libs/ui/util/src/index.ts"]
    },
    "typeRoots": ["types", "node_modules/@types"],
    "types": ["@emotion-types", "@window-types"]
  },
  "exclude": ["node_modules", "tmp"],
  "files": [],
  "include": []
}
```

type dir
![image](https://user-images.githubusercontent.com/13616435/180902679-76b7abd2-9864-46fd-b1c7-71159c61bb58.png)

[SO post](https://stackoverflow.com/questions/69403269/global-type-declarations-within-a-nx-and-nextjs-monorepo) - second answer

**Reply by DaSchTour** [reply]
I tried the same but it doesn't work.

I have in my tsconfig base

```
"typeRoots": [
      "types"
    ],
    "types": ["document"]
```

And then a types/document/index.d.ts with
```
interface Document {
  webkitFullscreenElement: Document['fullscreenElement'];
}
```

But `webkitFullscreenElement` is still not allowed on document.

**Reply by ALFmachine** [reply]
@DaSchTour apologies for the late reply. 

In newer versions of Nx we need to add the types to the `tsconfig.app.json` as well, I would try that as a next step.

**Reply by ste7en** [reply]
It's really annoying that each generated lib or app by Nx overrides the `types` tsconfig field

**Reply by ALFmachine** [reply]
I believe this is due to typescript not NX specifically but yeah, it would be better if that wasn't a required thing to do. positive side at least there's a way 😄

**Reply by knight10-ux** [reply]
> It's really annoying that each generated lib or app by Nx overrides the `types` tsconfig field

This could be solved with `compilerOptions` extends keyword available from typescript 5.0. There is discussion for this already https://github.com/nrwl/nx/discussions/15269

**Reply by vemundeldegard** [reply]
> > It's really annoying that each generated lib or app by Nx overrides the `types` tsconfig field
> 
> This could be solved with `compilerOptions` extends keyword available from typescript 5.0. There is discussion for this already #15269

Could you clarify if this would be working or not in 5.0? I'm a bit confused :D
